### PR TITLE
Add Vloop BC

### PR DESCRIPTION
**Pull request by theo-brown**
Adding Vloop boundary condition on the psi equation, so that the current can be evolved self-consistently given a rate of change of edge flux.

Tasks:
- [x] Add Vloop boundary condition
- [x] Add Vloop_LCFS to outputs
- [x] Support all methods of initialising psi
- [x] Unit test
- [x] Integration test

### CODE REVIEW

**Comment by theo-brown** [comment]
I have not yet written a test case for this, as I wanted to run the pattern past one of the team first. Let me know your initial thoughts and then I can continue!

**Review by jcitrin — COMMENTED** [review]
Some general thoughts.

1. What's missing is a consideration of how to initialize psi to begin with, for all cases. If Ip is None, then how do we set the right boundary condition at the first timestep, for cases where we initialize based on plasma current? Recall that there are several options currently (which can be further extended later):

i) profile_conditions['initial_psi_from_j'] = False, i.e. Initialize psi based on the input geometry file. Two suboptions:
a)  geometry['Ip_from_parameters'] = False. 
Here's it's clear what to do, psi_right = psi_geometry[-1]
b) geometry['Ip_from_parameters'] = True. For this we need Ip to rescale psi, so if Ip=None we have a problem

ii) profile_conditions['initial_psi_from_j'] = True, i.e. Initialize psi based on the "nu" current profile model. This needs normalization to a total current, i.e. Ip. So if Ip=None we have a problem. Note that at the initial condition, the Vloop_bound_right needs to be such that Psi.face_grad()[-1] is consistent with Ip, according to the same formula as the standard boundary condition. See fvm.cell_variable.face_grad, that face_grad()[-1] is just a standard derivative between the dx/2 distance between the last cell point and the rightmost face.

To solve this, I recommend changing the name of Ip to Ip_input, and for cases where the Vloop boundary condition is used, then it's used (if necessary) to initialize psi. That means that Vloop_bound_right and Ip can coexist. The pattern should be a bit different, e.g. an extra boolean "use_Vloop_boundary_condition" or something like that. Note that Ip_input can still be a TimeInterpolatedInput, but for the Vloop case only the initial condition at t=t_initial will be used.

2. We need to adjust to the Vloop boundary case where profile_conditions['Ip_input'] is not actually total current, but rather the initial condition. For example, the output Ip is taken from there, and also in plotruns_lib.PlotData.i_total. For the code to work with both boundary condition cases, we should take Ip (total current) from elsewhere.  A candidate could be to extend physics.calc_jtot_from_psi to output I_tot as well (can change the name to Ip_profile), and then save this quantity in CoreProfiles.Currents . Total current is then Ip_profile[-1]. This can be a separate PR. This way, we also get information on the total current evolution in the Vloop boundary condition case.

File: torax/core_profile_setters.py [hunk 1]
```diff
@@ -579,18 +581,29 @@ def _update_psi_from_j(
...
 
+def _psi_value_constraint_from_Vloop(
+  dynamic_runtime_params_slice_t: runtime_params_slice.DynamicRuntimeParamsSlice,
+  core_profiles_t_minus_dt: state.CoreProfiles,
+  geo: Geometry,
+  dt: jax.Array,
+) -> jax.Array:
+  """Calculates the value constraint on the poloidal flux (psi) from Vloop."""
+  return (
+    core_profiles_t_minus_dt.psi.face_value(geo.rho_b)
```
Comment (jcitrin) [inline torax/core_profile_setters.py:604]: should be face_value()[-1]

**Comment by theo-brown** [comment]
Great points, and this is exactly why I wanted to run it through you! I think because for the type of simulation I've learnt how to do we always prescribe psi from a geometry I hadn't really thought about the consequences for any situation other than 1ia).

Noting down for myself - I need to consider the cases:

|    | Psi from... | Ip |
|---|----|---|
|   1ai | Geo | Self-consistent |
| 1aii  | Geo | Prescribed | 
| 2 | Initial j profile | Prescribed |

and their interactions with the Vloop BCs

**Comment by jcitrin** [comment]
#492 was done with this PR in mind

**Comment by theo-brown** [comment]
@jcitrin did #492 tackle all of the cases where `Ip_tot` was used as the plasma current internally, or was it just the writing to output?

**Comment by jcitrin** [comment]
> @jcitrin did #492 tackle all of the cases where `Ip_tot` was used as the plasma current internally, or was it just the writing to output?

Only the writing to output.

**Comment by theo-brown** [comment]
I've been working on incoporating the Vloop BC via [`compute_boundary_conditions`](
https://github.com/theo-brown/torax/blob/3b7f33c963cae820121485b2b52b26a856ef728c/torax/core_profile_setters.py#L969-L989).
The basic logic is (pseudocode):
```
if Vloop_bound_right is given:
   psi right BC = value constraint from Vloop
else:
   psi right BC = grad constraint from Ip_tot
```
In this case, I've attempted to sidestep the initialisation problem by having a spatial-derivative BC for the first timestep and a value BC for subsequent timesteps.

I've found that my current method results in an inconsistency with the setting of psi vs psidot.
The psi that is written to the output **(`core_profiles/psi`) is consistent** with `Vloop_bound_right`. In this case, `Vloop_bound_right = 0`, so the time derivative of psi at the LCFS is 0.
However, **`core_profiles/psidot` is not consistent** with this (see graph).

psidot

From the code, it appears that within `sim.py`, `psidot` is calculated using the function from `ohmic_heat_source.py`:

https://github.com/google-deepmind/torax/blob/c86055723b03e6bbd39a8d982e281c9ebfb92b75/torax/sim.py#L1568-L1583

https://github.com/google-deepmind/torax/blob/c86055723b03e6bbd39a8d982e281c9ebfb92b75/torax/sources/ohmic_heat_source.py#L43-L138

I don't understand yet what's going on in `calc_psidot`, or how it interacts with the `compute_boundary_conditions`. 
Looking at `calc_coeffs` it looks like the same logic is indeed happening there, which is probably the most significant location out of all of them. This gives me the impression that **modifying `compute_boundary_conditions` is insufficient to implement this functionality**. @jcitrin do you have any insights on whether that's the case, or whether I've just missed something? Thanks in advance for your help!

**Comment by jcitrin** [comment]
calc_psidot literally calculates the entire RHS of the governing psi equation.

So \frac{\partial \psi}{\partial t} = psidot 

First thing, is that psidot is on the _cell_ grid. Therefore psidot[:, -1] does not correspond to the LCFS, so you shouldn't expect equivalence.

For a basic sanity check of psidot, you can try setting 100% non-inductive current in one of the configs and see psidot become both flat (stationary-state current profile) and zero (no inductive current that shifts psi in time). See below for iterhybrid_predictor_corrector when turning off bootstrap current and doing the following for generic_current_source

        'generic_current_source': {
            # total "external" current fraction
            'fext': 1.0,
            # width of "external" Gaussian current profile (normalized radial
            # coordinate)
            'wext': 0.4,
            # radius of "external" Gaussian current profile (normalized radial
            # coordinate)
            'rext': 0.0,
        },

plot below: psidot[-1] is y-axis, time is x-axis
image

Therefore what's happening to psi and psidot at the LCFS is currently not in the output, and wouldn't even be accessible from the calc_psidot function, because that doesn't have any info on the boundary conditions.

It would indeed be useful to include Vloop_LCFS as a core_profile output. It can be calculated using CellVariable methods based on psi from core_profiles_t and core_profiles_t_plus_dt. If not setting Vloop_LCFS as a boundary condition and just calculating it based on the profiles with fixed Ip, then the initial time Vloop will be undefined.

Regarding boundary conditions, I recommend for consistency, setting the Dirichlet boundary conditions throughout, when the Vloop option is used. On initialization, you must make it consistent with the initial total Ip, by calculating the necessary gradient  from the last cell gridpoint to the last face gridpoint, and doing linear extrapolation.

Happy to discuss over a call on Mon or Tue if needed.

**Comment by theo-brown** [comment]
> It would indeed be useful to include Vloop_LCFS as a core_profile output. 
Just to check on this, do you mean a property of `state.PostProcessedOutputs`? Or in `state.CoreProfiles`? 
In OMAS, it's in `core_profiles.global_quantities.v_loop`.

**Comment by jcitrin** [comment]
> Just to check on this, do you mean a property of state.PostProcessedOutputs? Or in state.CoreProfiles?
In OMAS, it's in core_profiles.global_quantities.v_loop.

Makes most sense in to be in CoreProfiles

**Comment by theo-brown** [comment]
This PR is now ready for review :)

## Summary of changes
- Added new BC for the `psi` equation, `ProfileConditions.vloop_lcfs`.
  - Can be combined with any of the psi initialisation methods:
    - Prescribed psi + Ip_tot rescaling
    - Psi from geometry file
    - Psi from geometry file + Ip_tot rescaling
    - Psi from nu formula + Ip_tot rescaling
- Added `vloop_lcfs` to `state.CoreProfiles`, which is a direct copy from `psidot.face_value()[-1]`.
- Added `psi_right_bc` and `vloop_lcfs` to output.

## Outstanding questions/bugs
- The options involving psi from geometry file are not _unit_ tested, as the other unit tests of this type avoided reading in a geometry file. They are, however, tested by _integration_ tests.
- The prescribed psi case is not _integration_ tested, as the other tests for the psi equation seemed to all use the nu formula.
-  As `psi_right_bc` can be None, this required changing one of the unit tests so that `assertequal` allowed `None == None`.  **I'm concerned that this is slightly bad practice.**   
- It also results in a warning when constructing the output, as the output involves converting a `jnp.array([None, ...]) -> jnp.array([jnp.nan, ....])`, which will be deprecated in future. For ease of comparing `.nc` files and Xarray Datasets, I thought it was better to have the key `psi_right_bc` _always present_ but sometimes `nan`, rather than having only _sometimes present_. I'm open to other opinions on the matter!
- I think that these changes mean that at least one profile for the first timestep has a nan in (q at edge = inf). This breaks the plotting script. If I have time this week I'll track through and find why this happens.

**Comment by theo-brown** [comment]
Not sure why this failed copybara - is it because it's behind main so there are conflicts?

**Comment by jcitrin** [comment]
Indeed, it's due to merge conflicts. Strongly recommend merging main and syncing to HEAD. It may get a bit hairy, sorry about that. 😬 

And apologies again for the delays in review. Will get on it tonight

**Comment by theo-brown** [comment]
OK! I'll give it a go this morning.
Don't worry about the delay - the new features are super exciting, I was more than happy to wait!

**Comment by theo-brown** [comment]
I've rebased onto main, not sure why it's still failing.

**Comment by jcitrin** [comment]
> I've rebased onto main, not sure why it's still failing.
Yeah, weird. Maybe it's due to merge conflicts in the .nc files which it doesn't know how to resolve. We can resolve them manually when we do the manual import and internal review, so I wouldn't worry about it. As long as it's rebased to main and the actual tests are passing, should be fine.

File: torax/core_profile_setters.py [hunk 2]
```diff
@@ -923,8 +1011,10 @@ def get_update(x_new, var):
...
 
 def compute_boundary_conditions(
+    dt: jax.Array,
     static_runtime_params_slice: runtime_params_slice.StaticRuntimeParamsSlice,
-    dynamic_runtime_params_slice: runtime_params_slice.DynamicRuntimeParamsSlice,
+    dynamic_runtime_params_slice_t: runtime_params_slice.DynamicRuntimeParamsSlice,
+    core_profiles_t_minus_dt: state.CoreProfiles,
     geo: geometry.Geometry,
 ) -> dict[str, dict[str, jax.Array | None]]:
   """Computes boundary conditions for time t and returns updates to State.
```
Comment (jcitrin) [inline torax/core_profile_setters.py:1020]: In light of above, can change this to "...for time t_plus_dt and returns..."

File: torax/core_profile_setters.py [hunk 3]
```diff
@@ -923,8 +1011,10 @@ def get_update(x_new, var):
 
 
 def compute_boundary_conditions(
+    dt: jax.Array,
```
Comment (jcitrin) [inline torax/core_profile_setters.py:1014]: see also comment where in step_function where compute_boundary_conditions is called.

I think it's more consistent to change the namings and to do the following:

1. change the function name to "compute_boundary_conditions_for_t_plus_dt"  (this is its typical use-case)

2. Have the following input arguments:

dt
static_runtimes_params_slice
dynamic_runtime_params_slice_t
dynamic_runtime_params_slice_t_plus_dt
core_profiles_t

Then, for all the bc setting for e.g. Ti_bound_right and Te_bound_right, etc., and _calculate_psi_grad_constraint_from_Ip_tot,  where you changed 
dynamic_runtime_params_slice --> dynamic_runtime_params_slice_t,  this should now be
dynamic_runtime_params_slice --> dynamic_runtime_params_slice_t_plus_dt 

And for _psi_value_constraint_from_vloop, the way it is now is hard-coded it to be fully implicit, i.e. Vloop from time t_plus_dt is what's setting psi_lcfs_t_plus_dt. To be consistent with the TORAX theta method, should be a theta-weighted average, i.e.  theta*Vloop_t + (1-theta)*Vloop_t_plus_dt.   

theta can be found in the static_runtime_params, and dynamic_runtime_params_t should then also be passed through (which is easy since it's available in the calling functions).

File: torax/orchestration/step_function.py [hunk 4]
```diff
@@ -747,9 +751,11 @@ def _provide_core_profiles_t_plus_dt(
   """Provides state at t_plus_dt with new boundary conditions and prescribed profiles."""
   updated_boundary_conditions = (
       core_profile_setters.compute_boundary_conditions(
-          static_runtime_params_slice,
```
Comment (jcitrin) [inline torax/orchestration/step_function.py:750]: see comment in `compute_boundary_conditions` regarding a suggestion on naming and interpretation consistencies, and also passing in both dynamic_runtime_params_t and dynamic_runtime_params_t_plus_dt for enabling the theta method in the psi_lcfs calculation

File: torax/core_profile_setters.py [hunk 5]
```diff
@@ -784,7 +868,11 @@ def initial_core_profiles(
           source_models,
       ),
   )
-  core_profiles = dataclasses.replace(core_profiles, psidot=psidot)
+  core_profiles = dataclasses.replace(
+    core_profiles,
+    psidot=psidot,
+    vloop_lcfs=psidot.face_value()[-1],
```
Comment (jcitrin) [inline torax/core_profile_setters.py:874]: I think this might be a bug for the initialization?

To be honest I'm not sure that psidot should really be a CellVariable (my mistake). It's only calculated for the ohmic heat source and is on the cell grid. The "boundary conditions" of the CellVariable are initialized to zero and never updated. So doing `psidot.face_value()[-1]` here will just set it to zero and overwrite any correct initialization from vloop_lcfs

File: torax/core_profile_setters.py [hunk 6]
```diff
@@ -34,7 +34,7 @@
 from torax.config import runtime_params_slice
 from torax.fvm import cell_variable
 from torax.geometry import geometry
-from torax.sources import ohmic_heat_source
+from torax.sources import ohmic_heat_source
```
[the remark this answers is not in the document]
Comment (jcitrin) [reply torax/core_profile_setters.py:37]: @Nush395 , @tamaranorman , @sbodenstein : any thoughts here? I'm less knowledgeable about this

[the remark this answers is not in the document]
Comment (tamaranorman) [reply torax/core_profile_setters.py:37]: So apparently  isort+black should do it

The other thing you can do is run pylint and optionally with this which will give additional warnings etc

[the remark this answers is not in the document]
Comment (tamaranorman) [reply torax/core_profile_setters.py:37]: Does pyink not have an autoformatter implementation you can use

As it seems that is the best way of working with it if possible

[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/core_profile_setters.py:37]: Yeah sorry for the wasted time, this is clearly a user-side problem to do with me not reading the pyink README properly. 

I'll fix it up my end and then should be free from such small linting errors in future.

File: torax/orchestration/step_function.py (hunk 4 above)
Comment (theo-brown) [reply torax/orchestration/step_function.py:750]: Resolved in [d80f1f6](https://github.com/google-deepmind/torax/pull/456/commits/d80f1f6c54d98621589b9977a953ad58ddb17065)

File: torax/core_profile_setters.py (hunk 3 above)
Comment (theo-brown) [reply torax/core_profile_setters.py:1025]: Resolved in [d80f1f6](https://github.com/google-deepmind/torax/pull/456/commits/d80f1f6c54d98621589b9977a953ad58ddb17065)

File: torax/core_profile_setters.py (hunk 2 above)
Comment (theo-brown) [reply torax/core_profile_setters.py]: Resolved in [d80f1f6](https://github.com/google-deepmind/torax/pull/456/commits/d80f1f6c54d98621589b9977a953ad58ddb17065)

File: torax/core_profile_setters.py [hunk 7]
```diff
@@ -577,6 +590,19 @@ def _calculate_psi_grad_constraint(
   )
 
 
+def _psi_value_constraint_from_vloop(
+  dt: jax.Array,
+  dynamic_runtime_params_slice_t: runtime_params_slice.DynamicRuntimeParamsSlice,
+  core_profiles_t_minus_dt: state.CoreProfiles,
+  geo: geometry.Geometry,
+) -> jax.Array:
+  """Calculates the value constraint on the poloidal flux (psi) from loop voltage at LCFS."""
```
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/core_profile_setters.py]: Resolved in [d80f1f6](https://github.com/google-deepmind/torax/pull/456/commits/d80f1f6c54d98621589b9977a953ad58ddb17065)

File: torax/tests/test_data/test_psichease_ip_parameters_vloop.py
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/tests/test_data/test_psichease_ip_parameters_vloop.py:1]: Fixed in [72a2d80](https://github.com/google-deepmind/torax/pull/456/commits/72a2d8034d2ec156a558c3009b2adfc3baca693f)

File: torax/tests/test_data/test_psichease_ip_parameters_prescribed_jtot_vloop.py
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/tests/test_data/test_psichease_ip_parameters_prescribed_jtot_vloop.py:1]: Fixed in [72a2d80](https://github.com/google-deepmind/torax/pull/456/commits/72a2d8034d2ec156a558c3009b2adfc3baca693f)

File: torax/tests/test_data/test_psichease_ip_chease_vloop.py
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/tests/test_data/test_psichease_ip_chease_vloop.py:1]: [72a2d80](https://github.com/google-deepmind/torax/pull/456/commits/72a2d8034d2ec156a558c3009b2adfc3baca693f)

File: torax/core_profile_setters.py (hunk 5 above)
Comment (theo-brown) [reply torax/core_profile_setters.py]: You're right this is overwriting the initialisation from lines 831-835, which isn't the right thing to do.

I think because the default branch in `CellVariable.face_value()` just copies the value at the cell centre to the edge ([see cell_variable.py](https://github.com/google-deepmind/torax/blob/6c068486e868e72a3d936c737fc9ae86a914ea7c/torax/fvm/cell_variable.py#L214-L221)), this doesn't set it to 0 but instead the cell value psidot[-1] - I've only just realised this.

Maybe on Thursday we could talk about what the right thing is to do here. I think revisiting the vloop initialisation more widely is necessary, to be honest, and it would be really great to discuss.

I can switch psidot to an array rather than a CellVariable as a drive-by if that would be useful. It would require changing the type in the CoreProfiles, not sure what else.

File: torax/output.py [hunk 8]
```diff
@@ -202,9 +204,9 @@ def __init__(
     post_processed_output = [
         state.post_processed_outputs for state in sim_outputs.sim_history
     ]
-    stack = lambda *ys: jnp.stack(ys)
+    stack = lambda *ys: jnp.stack(jnp.array(ys))
     self.core_profiles: state.CoreProfiles = jax.tree_util.tree_map(
-        stack, *core_profiles
+        stack, *core_profiles, is_leaf=lambda x: x is None,
```
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/output.py:209]: This is a workaround to catch the fact that either psi_right_bc or psi_right_grad_bc is None (although I'll be honest I don't quite understand how it works)
In hindsight, I need a better fix for this because it's also blocking the plotting.

**Comment by theo-brown** [comment]
I'm fixing some of the merge conflicts. Is it alright to revert some of 9c76cc0 (line 570 in `core_profile_setters`, as _update_psi_from_j now needs to be able to select between branches based on the BC set in the runtime pararms slice?

**Comment by jcitrin** [comment]
Is there another way, like by extending the input arguments with
information on which branch to take, or having a different function for the
different branch?

File: torax/core_profile_setters.py [hunk 9]
```diff
@@ -604,16 +631,39 @@ def _init_psi_and_current(
...
+  # Case 1: retrieving psi from the profile conditions, using the prescribed profile and Ip_tot
   if dynamic_runtime_params_slice.profile_conditions.psi is not None:
+    # Calculate the dpsi/drho necessary to achieve the given Ip_tot
+    dpsi_drhonorm_edge = _calculate_psi_grad_constraint_from_Ip_tot(
+        dynamic_runtime_params_slice,
+        geo,
+    )
+
+    # Set the BCs to ensure the correct Ip_tot
+    if use_vloop_as_boundary_condition:
```
Comment (tamaranorman) [inline torax/core_profile_setters.py:647]: I know this was true already but there is now a lot of repeated/confusing code here

Can we maybe look tomorrow at a way to refactor/filter this a bit

And also apologies in advance that there may be some more merge conflicts here, we have obviously all decided to fix the same bit of code

Comment (theo-brown) [reply torax/core_profile_setters.py:647]: Definitely agree, I'm not particularly happy with how this function has ended up looking. 
I'm currently trying to fix up the unit tests, and then will try and achieve ~zero merge conflicts before tomorrow so that we can focus on refactoring.

File: torax/output.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/output.py:209]: I've changed this so that there is a separate flag for `use_vloop_lcfs_boundary_condition`. Should render this change unnecessary, propagating it through now.

**Comment by theo-brown** [comment]
> Is there another way, like by extending the input arguments with information on which branch to take, or having a different function for the different branch?

Done in [4d55436](https://github.com/google-deepmind/torax/pull/456/commits/4d554362aaaa61139f2b7102895e29791fc4092a)

File: torax/output.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/output.py]: Fixed in [599a2ee](https://github.com/google-deepmind/torax/pull/456/commits/599a2eee61d3b3734032df08772b253ccce087d5).

**Comment by theo-brown** [comment]
Behaviour of psidot, psi edge, and Ip in updated sims ([986f3ec](https://github.com/google-deepmind/torax/pull/456/commits/986f3ec6e9d1cab0a40497ebe5df6b13b6cbff85))

image
image
image

File: torax/tests/sim.py [hunk 10]
```diff
@@ -447,6 +447,30 @@ class SimTest(sim_test_case.SimTestCase):
           _ALL_PROFILES,
           0,
       ),
+      # Tests current diffusion with vloop BC.
```
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/tests/sim.py:457]: Added screenshots below

File: torax/config/profile_conditions.py [hunk 11]
```diff
@@ -208,6 +216,8 @@ class DynamicProfileConditions:
   """Prescribed values and boundary conditions for the core profiles."""
 
   Ip_tot: array_typing.ScalarFloat
+  vloop_lcfs: array_typing.ScalarFloat
+  use_vloop_lcfs_boundary_condition: bool
```
[the remark this answers is not in the document]
Comment (theo-brown) [reply torax/config/profile_conditions.py:220]: Deferred to future issue/PR

**Comment by jcitrin** [comment]
Things look even more sane after introducing psi_from_Ip_face and making that the right BC for Case 2.

Here's the new test-case with Ip_from_parameters

image


And here with Ip from CHEASE

image

Main point to notice is the Ip time series. In each case it starts from where expected (15MA, ~11.5MA  respectively), before Vloop begins to modify it. Previously I think we were effectively setting the initial current to zero by setting the BC as the last point from the cell grid, and hence there was a current perturbation in the first timesteps

**Comment by theo-brown** [comment]
Huge thanks to @jcitrin, @tamaranorman, and @Nush395 for your help in getting this in - especially for your suggestions and time yesterday, as well as extra work put in after it went to internal review :rocket:

**Comment by jcitrin** [comment]
It's a great feature, thanks for your efforts in getting this in! Looking forward to seeing its use in STEP simulations
